File: src/main/java/view25d/view/PatchView.java
package view25d.view;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.event.WindowEvent;
import java.awt.event.WindowListener;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import com.jogamp.opengl.awt.GLCanvas;
import com.jogamp.opengl.GLCapabilities;
import com.jogamp.opengl.GLProfile;
import javax.swing.JFrame;

import org.nlogo.agent.AgentSet;
import org.nlogo.agent.Turtle;
import org.nlogo.api.Agent;
import org.nlogo.api.AgentException;
import org.nlogo.api.Context;
import org.nlogo.api.Link;
import org.nlogo.api.Patch;
import org.nlogo.api.AnonymousReporter;
import org.nlogo.app.App;

import view25d.View25DExtension;
import view25d.view.gl.PatchGL;

public class PatchView extends VarviewWindow {

    private static final long serialVersionUID = 1L;

    private PatchGL glManager;
    public MouseableGLWindow getGLWindow() { return glManager; }
    private AnonymousReporter reporter;

    public PatchViewOptions viewOptions;

    public PatchValue[][] reporterValueMatrix;

    public boolean doingTurtles = false;
    public ArrayList<TurtleValue> turtleValues = new ArrayList<TurtleValue>();
    public ArrayList<LinkValue> linkValues;

    public PatchView(String title, AnonymousReporter rt) {
        super(title);
        reporter = rt;
    }

    //called by class that instantiates PatchView, after constructor.
    public void postConstructor() {
        setupUI();
        this.setDefaultCloseOperation( JFrame.DISPOSE_ON_CLOSE );
        final String myTitle = this.getTitle();
        //window listener to ensure that the HashMap gets cleaned up.
        this.addWindowListener(new WindowListener() {
            public void windowActivated(WindowEvent e) {}
            public void windowClosed(WindowEvent e) {}
            public void windowClosing(WindowEvent e) { View25DExtension.removePatchWindowWithTitle(myTitle); }
            public void windowDeactivated(WindowEvent e) {}
            public void windowDeiconified(WindowEvent e) {}
            public void windowIconified(WindowEvent e) {}
            public void windowOpened(WindowEvent e) {}
        });
        GLCapabilities glCapabilities = new GLCapabilities(GLProfile.get(GLProfile.GL2));
        GLCanvas glCanvas =  new GLCanvas(glCapabilities);

        glManager = new PatchGL(this);

        glCanvas.addGLEventListener(glManager);
        glCanvas.addMouseListener(glManager);
        glCanvas.addMouseMotionListener(glManager);
        glCanvas.addMouseWheelListener(glManager);

        glManager.setCanvas( glCanvas );
        mainPanel.add(glCanvas, BorderLayout.CENTER);

        viewOptions = new PatchViewOptions( this, true, false, false);
        mainPanel.add(viewOptions, BorderLayout.NORTH);

        setSize(new Dimension(600, 650));
        xExpand = 0;
        yExpand = 50;
        /*      this.addComponentListener( new ComponentAdapter() {
                public void componentResized(ComponentEvent e) {
                applySquareConstraint(xExpand, yExpand, mainPanel);
                }
                });
                */
        centerWindow(this);
    }



    public void setupForRendering(  int wWidth, int wHeight, int minX, int maxX, int minY, int maxY ) {
        worldWidth = wWidth;
        worldHeight = wHeight;
        minPxcor = minX;
        minPycor = minY;
        maxPxcor = maxX;
        maxPycor = maxY;
        reporterValueMatrix = new PatchValue[worldWidth][worldHeight];
    }


    public void updateValueArray(Context context) {
        for (int i = 0; i< worldWidth; i++){
            for (int j = 0;j< worldHeight; j++) {
                try {
                    Patch patch = App.app().workspace().world().getPatchAt(i + minPxcor, j + minPycor);
                    Color c = org.nlogo.api.Color.getColor(patch.pcolor());
                    double val = (Double)reporter.report(context, new Object[]{patch});
                    reporterValueMatrix[i][j] = new PatchValue(c, val);
                } catch (AgentException e) {
                    e.printStackTrace();
                }
            }
        }

        ArrayList<TurtleValue> temp = new ArrayList<TurtleValue>();
        ArrayList<LinkValue> temp2 = new ArrayList<LinkValue>();
        if (doingTurtles) {
            AgentSet as = App.app().workspace().world().turtles();
            for (Agent a : as.agents() ) {
                Turtle t = (Turtle)a;
                double val = (Double)reporter.report(context, new Object[]{t});
                Color c = org.nlogo.api.Color.getColor(t.color());
                double stemColor = 4.5f;  // This will be ignored no stem in PatchView
                TurtleValue tv = new TurtleValue( t.shape(), c, t.size(), t.xcor(), t.ycor(), val, stemColor );
                tv.setHeading(t.heading());
                temp.add(tv);
            }
            // Get Set of Links associated with the Turtles
            Set<Link> linkSet = TurtleView.getLinkSetFromTurtleSet(as);
            for (Link link : linkSet) {

                Turtle end1 = (Turtle)link.end1();
                Turtle end2 = (Turtle)link.end2();

                Color c = org.nlogo.api.Color.getColor(link.color());
                double zcor1 = (Double)reporter.report(context, new Object[]{end1});
                double zcor2 = (Double)reporter.report(context, new Object[]{end2});

                LinkValue lv = new LinkValue(link.shape(), c, link.lineThickness(),
                                             end1.xcor(), end1.ycor(), zcor1,
                                             end2.xcor(), end2.ycor(), zcor2);
                temp2.add(lv);
            }
        }
        //copy over the new values in an atomic step to avoid concurrent access
        turtleValues = temp;
        linkValues = temp2;
    }

    @Override
    public void manuallyRefreshReporterView(Context context) {
        updateValueArray( context );
        glManager.repaintCanvas();
    }

    public void refresh() {
        glManager.repaintCanvas();
    }

    @Override
    public void resetPerspective() {
        glManager.observer.goHome(this);
        glManager.repaintCanvas();
    }

    public void zoomZby( double change ) {
        if (-change < zScale) {
            zScale += change;
        }
        glManager.repaintCanvas();
    }

}

File: src/main/scala/view/PatchView.scala
package view25d.view

import com.jogamp.opengl.{ GLCapabilities, GLProfile }
import com.jogamp.opengl.awt.GLCanvas

import java.awt.{ BorderLayout, Dimension }
import java.awt.event.{ WindowAdapter, WindowEvent }
import javax.swing.WindowConstants

import org.nlogo.agent.Turtle
import org.nlogo.api.{ AgentException, AnonymousReporter, Color, Context }
import org.nlogo.app.App
import org.nlogo.theme.ThemeSync

import scala.jdk.CollectionConverters.IterableHasAsScala

import view25d.View25DExtension
import view25d.view.gl.PatchGL

class PatchView(title: String, reporter: AnonymousReporter) extends VarviewWindow(title) with ThemeSync {
  private val glManager = new PatchGL(this)

  def getGLWindow: MouseableGLWindow =
    glManager

  var viewOptions = new PatchViewOptions(this, true, false, false, true, false)

  var reporterValueMatrix = Array[Array[PatchValue]]()

  var doingTurtles = false

  var turtleValues = Array[TurtleValue]()
  var linkValues = Array[LinkValue]()

  setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE)

  // window listener to ensure that the Map gets cleaned up.
  addWindowListener(new WindowAdapter {
    override def windowClosing(e: WindowEvent): Unit = {
      View25DExtension.removePatchWindowWithTitle(title)
    }
  })

  val glCanvas =  new GLCanvas(new GLCapabilities(GLProfile.get(GLProfile.GL2)))

  glCanvas.addGLEventListener(glManager);
  glCanvas.addMouseListener(glManager);
  glCanvas.addMouseMotionListener(glManager);
  glCanvas.addMouseWheelListener(glManager);

  glManager.setCanvas(glCanvas)

  mainPanel.add(glCanvas, BorderLayout.CENTER)
  mainPanel.add(viewOptions, BorderLayout.NORTH)

  setSize(new Dimension(600, 650))

  centerWindow(this)

  syncTheme()

  def setupForRendering(wWidth: Int, wHeight: Int, minX: Int, maxX: Int, minY: Int, maxY: Int): Unit = {
    worldWidth = wWidth
    worldHeight = wHeight
    minPxcor = minX
    minPycor = minY
    maxPxcor = maxX
    maxPycor = maxY

    reporterValueMatrix = Array.fill(worldWidth) { new Array[PatchValue](worldHeight) }
  }

  def updateValueArray(context: Context): Unit = {
    for (i <- 0 until worldWidth){
      for (j <- 0 until worldHeight) {
        try {
          val patch = App.app.workspace.world.getPatchAt(i + minPxcor, j + minPycor)
          val c = Color.getColor(patch.pcolor)

          reporterValueMatrix(i)(j) = new PatchValue(c, reporter.report(context, Array(patch)).asInstanceOf[Double])
        }

        catch {
          case e: AgentException => e.printStackTrace()
        }
      }
    }

    if (doingTurtles) {
      val as = App.app.workspace.world.turtles

      turtleValues = as.agents.asScala.toArray.collect {
        case turtle: Turtle if !turtle.hidden =>
          // stemColor will be ignored, no stem in PatchView
          TurtleValue(turtle.shape, Color.getColor(turtle.color), turtle.xcor, turtle.ycor, turtle.size,
                      reporter.report(context, Array(turtle)).asInstanceOf[Double], turtle.heading, 4.5f)
      }

      // Get Set of Links associated with the Turtles
      linkValues = getLinkSetFromTurtleSet(as).toArray.map(link => {
        val end1 = link.end1
        val end2 = link.end2

        LinkValue(link.shape, Color.getColor(link.color), link.lineThickness, end1.xcor, end1.ycor,
                  reporter.report(context, Array(end1)).asInstanceOf[Double], end2.xcor, end2.ycor,
                  reporter.report(context, Array(end2)).asInstanceOf[Double])
      })
    }

    else {
      turtleValues = Array()
      linkValues = Array()
    }
  }

  override def manuallyRefreshReporterView(context: Context): Unit = {
    updateValueArray(context)
    glManager.repaintCanvas()
  }

  def refresh(): Unit = {
    glManager.repaintCanvas()
  }

  def updateTurtleShapes(): Unit = {
    glManager.updateTurtleDisplayList()
    glManager.repaintCanvas()
  }

  override def resetPerspective(): Unit = {
    glManager.observer.goHome(this)
    glManager.repaintCanvas()
  }

  def zoomZby(change: Double): Unit = {
    if (-change < zScale)
      zScale += change

    glManager.repaintCanvas()
  }

  def syncTheme(): Unit = {
    dashboard.syncTheme()
    viewOptions.syncTheme()
  }
}
